File: heritage.java
// ---------------------------------------------------------------------------
// TP4 : Code hérité. VOLONTAIREMENT non testable.
//
// Ne cherchez pas l'erreur : il n'y en a pas. Le problème n'est pas que ce code
// soit faux, c'est qu'il est IMPOSSIBLE à tester unitairement en l'état.
// Votre premier travail est de comprendre pourquoi, avant de toucher à quoi que
// ce soit. Lisez tp/tp4-compte-bancaire.md.
// ---------------------------------------------------------------------------

import java.util.*;

public class CompteService {

    private static final int DECOUVERT_AUTORISE = -500;
    private static final int PLAFOND_DEPOT      = 10_000;

    public void debiter(String iban, int montant) {
        Compte compte = new BaseDeDonnees().charger(iban);
        int nouveauSolde = compte.solde() - montant;

        if (nouveauSolde < DECOUVERT_AUTORISE) {
            throw new OperationRefusee("Découvert autorisé dépassé");
        }

        compte.fixerSolde(nouveauSolde);
        new BaseDeDonnees().sauver(compte);
        new BaseDeDonnees().journaliser(new Transaction(iban, -montant, new Date()));

        if (nouveauSolde < 0) {
            new SmtpSender().envoyer(
                compte.email(),
                "Votre compte est à découvert depuis le " + new Date());
        }
    }

    public void crediter(String iban, int montant) {
        Compte compte = new BaseDeDonnees().charger(iban);
        int nouveauSolde = compte.solde() + montant;

        if (nouveauSolde > PLAFOND_DEPOT) {
            throw new OperationRefusee("Plafond de dépôt dépassé");
        }

        compte.fixerSolde(nouveauSolde);
        new BaseDeDonnees().sauver(compte);
        new BaseDeDonnees().journaliser(new Transaction(iban, montant, new Date()));
    }

    public List<Transaction> releve(String iban) {
        List<Transaction> transactions = new BaseDeDonnees().transactionsDe(iban);
        transactions.sort(Comparator.comparing(Transaction::date));
        return transactions;
    }
}

// --- Infrastructure fournie : vous n'avez pas à la réécrire -----------------

class Compte {
    private final String iban;
    private final String email;
    private int solde;

    Compte(String iban, String email, int solde) {
        this.iban = iban; this.email = email; this.solde = solde;
    }
    String iban()  { return iban; }
    String email() { return email; }
    int solde()    { return solde; }
    void fixerSolde(int s) { this.solde = s; }
}

record Transaction(String iban, int montant, Date date) { }

class OperationRefusee extends RuntimeException {
    OperationRefusee(String message) { super(message); }
}

/** Se connecte à une vraie base PostgreSQL. Indisponible depuis un test. */
class BaseDeDonnees {
    Compte charger(String iban)                  { throw new UnsupportedOperationException("PostgreSQL"); }
    void   sauver(Compte compte)                 { throw new UnsupportedOperationException("PostgreSQL"); }
    void   journaliser(Transaction transaction)  { throw new UnsupportedOperationException("PostgreSQL"); }
    List<Transaction> transactionsDe(String iban){ throw new UnsupportedOperationException("PostgreSQL"); }
}

/** Envoie un VRAI mail. Chaque exécution de test enverrait un message réel. */
class SmtpSender {
    void envoyer(String destinataire, String message) {
        throw new UnsupportedOperationException("smtp.banque.fr:587");
    }
}

File: heritage.ts
// ---------------------------------------------------------------------------
// TP4 : Code hérité. VOLONTAIREMENT non testable.
//
// Ne cherchez pas l'erreur : il n'y en a pas. Le problème n'est pas que ce code
// soit faux, c'est qu'il est IMPOSSIBLE à tester unitairement en l'état.
// Votre premier travail est de comprendre pourquoi, avant de toucher à quoi que
// ce soit. Lisez tp/tp4-compte-bancaire.md.
// ---------------------------------------------------------------------------

const DECOUVERT_AUTORISE = -500;
const PLAFOND_DEPOT = 10_000;

export interface Horloge {
  maintenant(): Date;
}

export interface TransactionJournalisee {
  charger(iban: string): Compte;
  sauver(compte: Compte): void;
  journaliser(transaction: Transaction): void;
  listerTransactions(iban: string): Transaction[];
}

export interface NotificationService {
  envoyer(destinataire: string, message: string): void;
}

export class CompteService {
  constructor(private readonly horloge: Horloge,
    private readonly transactionJournalisee: TransactionJournalisee,
    private readonly notificationService: NotificationService) { }

  debiter(iban: string, montant: number): void {
    const compte = this.transactionJournalisee.charger(iban);
    const nouveauSolde = compte.solde - montant;

    if (nouveauSolde < DECOUVERT_AUTORISE) {
      throw new OperationRefusee("Découvert autorisé dépassé");
    }

    compte.solde = nouveauSolde;
    this.transactionJournalisee.sauver(compte);
    this.transactionJournalisee.journaliser(new Transaction(iban, -montant, this.horloge.maintenant()));

    if (nouveauSolde < 0) {
      this.notificationService.envoyer(
        compte.email,
        `Votre compte est à découvert depuis le ${this.horloge.maintenant().toISOString()}`,
      );
    }
  }

  crediter(iban: string, montant: number): void {
    const compte = this.transactionJournalisee.charger(iban);
    const nouveauSolde = compte.solde + montant;

    if (nouveauSolde > PLAFOND_DEPOT) {
      throw new OperationRefusee("Plafond de dépôt dépassé");
    }

    compte.solde = nouveauSolde;
    this.transactionJournalisee.sauver(compte);
    this.transactionJournalisee.journaliser(new Transaction(iban, montant, this.horloge.maintenant()));
  }

  releve(iban: string): Transaction[] {
    const transactions = this.transactionJournalisee.listerTransactions(iban);
    return transactions.sort((a, b) => a.date.getTime() - b.date.getTime());
  }
}

// --- Infrastructure fournie : vous n'avez pas à la réécrire -----------------

export class Compte {
  constructor(
    public readonly iban: string,
    public readonly email: string,
    public solde: number,
  ) {}
}

export class Transaction {
  constructor(
    public readonly iban: string,
    public readonly montant: number,
    public readonly date: Date,
  ) {}
}

export class OperationRefusee extends Error {}
